### PR TITLE
Fixed Remote Code Execution Vulnerability

### PR DESCRIPTION
**Pull request by mufeedvh**
:gear:  **Fix:**

The `PDFImage()` function doesn't validate the input of `pdfFilePath` input parameter. The `pdfFilePath` parameter accepts the path of the PDF file which is then used in a system command execution which when given without validation will result in a Remote Code Execution (RCE) Vulnerability.

What the fix does is: It validates the `pdfFilePath` input and checks for characters that can be used to concatenate any other system commands.

:question: **How:**

The mitigation is implemented with a Regex (```/;|&|`|\$|\(|\)|\|\||\||!|>|<|\?|\${/g```) check for invalid characters in **_index.js_**

:paperclip: **Proof of Concept (PoC):**

**poc.js** (put inside the **_node-pdf-image_** folder)

```
var PDFImage = require('.').PDFImage;

var pdfImage = new PDFImage('"; sleep 500 #"');
pdfImage.getInfo();
```

:fire: **Fix On Action:**

**OSX:**

    $ brew install imagemagick ghostscript poppler

**Linux (Ubuntu):**

    $ sudo apt-get install imagemagick ghostscript poppler-utils

**Run PoC To Test Fix:**

    $ npm install
    $ node poc.js

:heart:  **After Fix:**

node-pdf-image-fixed-poc

<hr>

### :v: **Fixed!**

<hr>

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -8,6 +8,12 @@ var util = require("util");
 var exec = require("child_process").exec;
 
 function PDFImage(pdfFilePath, options) {
+  // validating the file path for invalid characters to prevent remote code execution
+  if (/;|&|`|\$|\(|\)|\|\||\||!|>|<|\?|\${/g.test(pdfFilePath)) {
```
Comment (toufik-airane) [inline index.js]: The regexp isn't easily readable and so maintainable and especially two parts of the expression.
The first part is `|\|\||\|` and the second is `|\${`.

**Review by toufik-airane — COMMENTED** [review]
Dear @mufeedvh,

Thank you for your contribution.

The regexp isn't easily readable and so maintainable and especially two parts of the expression.
Can you ease the regular expression or at least document it?

Best regards.

**Review by adam-nygate — CHANGES_REQUESTED** [review]
Also, it looks like you're duplicating matches for `$` symbols and pipes (matching both `\|\|` and `\|`). Feel free to revise and then we can accept the PR 🙂

**Comment by mufeedvh** [comment]
Hi,

Check out the revised commit, I just cleaned up the Regex to ```/[`!@#$%^&*()_+\-=\[\]{};':"\\|,<>\/?~]/``` to filter out special characters except `.` to only accept filenames.

File: index.js [hunk 2]
```diff
@@ -8,6 +8,13 @@ var util = require("util");
 var exec = require("child_process").exec;
 
 function PDFImage(pdfFilePath, options) {
+  // validating the file path for invalid characters to prevent remote code execution
+  var filter_chars = /[`!@#$%^&*()_+\-=\[\]{};':"\\|,<>\/?~]/;
```
Comment (toufik-airane) [inline index.js]: Not all characters could trigger remote code execution.
You've blacklisted some harmful and probably legit one such as _ (underscore), ~ (tilde), etc.

I would only filter these characters, that have potential command shell feature:
;|`$()&<>

**Review by toufik-airane — CHANGES_REQUESTED** [review]
Dear @mufeedvh,

You are restricting too many characters. Some characters seem not related to a command shell feature.

File: index.js (hunk 2 above)
Comment (mufeedvh) [reply index.js]: Oh yeah didn't think about it. I did it in a hurry and was sleepy! 😅

**Comment by mufeedvh** [comment]
Done, fixed that! :+1:

**Review by adam-nygate — APPROVED** [review]
LGTM 👍

**Review by toufik-airane — APPROVED** [review]

**Review by toufik-airane — APPROVED** [review]

**Comment by huntr-helper** [comment]
Congratulations @mufeedvh - your fix has been selected! :tada:<br><br>Thanks for being part of the community & helping secure the world\'s open source code.<br>If you have any questions, please respond in the comments section. Your bounty is on its way - keep hunting!

**Pull request merged**